### PR TITLE
Configurable wait timeout for process to end and `-w SECONDS` CLI option

### PR DESCRIPTION
**Pull request by iandennismiller**
This pull request introduces a new configurable option, `-w SECONDS`, to control the maximum time earlyoom waits for a killed process to exit. The implementation includes updates to the CLI, input validation, documentation, and associated tests. This change allows users to adjust the wait timeout (default 10 seconds) to better handle processes that take longer to terminate.

**New feature: Configurable process wait timeout**

* Added a `-w SECONDS` command-line option to set the maximum wait time (in seconds) for a killed process to exit, with input validation and a maximum limit of 24 hours (`86400` seconds). [[1]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecL179-R184) [[2]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR282-R291) [[3]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR342)
* Updated the `poll_loop_args_t` struct and related logic to store and use the new `kill_wait_timeout_secs` value, defaulting to 10 seconds. [[1]](diffhunk://#diff-1713f949b66c0ce86f50a3021b07dd22ede1812572b989cf9d91ed9d143ef82aR36-R37) [[2]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR148)
* Modified the process waiting logic in `kill_wait` to use the user-specified timeout, with overflow protection for the poll loop. [[1]](diffhunk://#diff-0bc65e62b22f542a4d4a54ff38bc3238ec54565b476faceec08b5db98e8cc590L264-R278) [[2]](diffhunk://#diff-0bc65e62b22f542a4d4a54ff38bc3238ec54565b476faceec08b5db98e8cc590L306-R316)

**Documentation and testing**

* Updated the documentation (`MANPAGE.md`) to describe the new `-w` option and its usage.
* Added and updated CLI tests to verify correct handling of the `-w` flag, including invalid values and edge cases.

**Internal constants and options**

* Introduced a new constant for the maximum allowed wait timeout and updated the CLI options enum accordingly. [[1]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR36-R38) [[2]](diffhunk://#diff-a0cb465674c1b01a07d361f25a0ef2b0214b7dfe9412b7777f89add956da10ecR49)

### CODE REVIEW

**Comment by rfjakob** [comment]
Hi, what problem does this solve?

**Comment by iandennismiller** [comment]
The process I'm monitoring takes longer than 10 seconds to shut down; and if it is terminated forcefully, it results in database corruption.

I need to prevent earlyoom from re-issuing SIGTERM every 10 seconds. Eventually, sending enough SIGTERMs will cause the process to terminate early, which corrupts the database. So, after 100 seconds and the receipt of 10 SIGTERMs, it will exit immediately as if it were SIGKILLed.

With the flag in this PR, I am specifying `-w 300` to to throttle how quickly earlyoom sends SIGTERM. The desired outcome is that subsequent SIGTERMs are sent less frequently. In practice, I haven't seen the process take longer than 900 seconds to finish. So for me, a wait time of 5 minutes fits the time scale of my monitored process better than the 10 second interval.

**Comment by rfjakob** [comment]
So your process counts SIGTERMs? What process is this?

**Comment by iandennismiller** [comment]
go-ethereum

**Comment by rfjakob** [comment]
Wow, really https://github.com/ethereum/go-ethereum/blob/3eed0580d4dc40fb15a14e1b0e9fc6985fefbfd7/cmd/utils/cmd.go#L110

**Comment by iandennismiller** [comment]
> Wow, really https://github.com/ethereum/go-ethereum/blob/3eed0580d4dc40fb15a14e1b0e9fc6985fefbfd7/cmd/utils/cmd.go#L110

Exactly; a bit idiosyncratic but there are non-production situations where it's handy, I suppose.

But for my purposes, I need to work around the 10 SIGTERMS behavior because it's not what I intend.

File: main.c [hunk 1]
```diff
@@ -176,7 +181,7 @@ int main(int argc, char* argv[])
     meminfo_t m = parse_meminfo();
 
     int c;
-    const char* short_opt = "m:s:M:S:kingN:P:dvr:ph";
+    const char* short_opt = "m:s:M:S:kingN:P:dvr:phw:";
```
Comment (rfjakob) [inline main.c:198]: Please make this `--kill-wait` instead of `-w`.

File: main.c [hunk 2]
```diff
@@ -324,6 +339,7 @@ int main(int argc, char* argv[])
                 "                            to 0 to disable completely\n"
                 "  -p                        set niceness of earlyoom to -20 and oom_score_adj to\n"
                 "                            -100\n"
+                "  -w SECONDS                max seconds to wait for a process to die (default 10)\n"
```
Comment (rfjakob) [inline main.c:362]: --kill-wait

File: MANPAGE.md [hunk 3]
```diff
@@ -117,6 +117,12 @@ When earlyoom is run through its default systemd service, the `-p` switch doesn'
     OOMScoreAdjust=-100
     Nice=-20
 
+#### -w SECONDS
```
Comment (rfjakob) [inline MANPAGE.md:126]: --kill-wait

File: kill.c [hunk 4]
```diff
@@ -261,11 +261,21 @@ int kill_release(const pid_t pid, const int pidfd, const int sig)
...
 int kill_wait(const poll_loop_args_t* args, pid_t pid, int sig)
 {
     const unsigned poll_ms = 100;
+    // Calculate max_poll_iterations safely to avoid overflow
+    // With poll_ms = 100, we get 10 iterations per second
+    // Max safe value: UINT_MAX / 10 = ~429496729 seconds (~13.6 years)
+    unsigned max_poll_iterations;
+    if ((unsigned)args->kill_wait_timeout_secs > UINT_MAX / 10u) {
+        // Cap at maximum to avoid overflow (should never happen in practice)
+        max_poll_iterations = UINT_MAX;
```
Comment (rfjakob) [inline kill.c]: Cannot happen. Delete the check.

**Comment by iandennismiller** [comment]
I am going to change the sequencing of this pull request and re-submit a new one in a bit. The problem is that I merged to master and based the PR on that, when I should have based the PR on the pre-merge branch. It will be simpler for me to cherry-pick the changes into a new branch that's already synced to yours.

**Comment by iandennismiller** [comment]
Closing this PR.

Please see https://github.com/rfjakob/earlyoom/pull/366 for the replacement PR with a cleaner merge path.
